### ISSUE TITLE
[Critical] Design defect: endianness is not stored and can be arbitrary

### ISSUE THREAD

**Issue by KOLANICH** [huggingface/safetensors#75]
```python
import numpy as np
import safetensors.numpy

a = safetensors.numpy.save({"a": np.array(range(6), dtype='>u4')})
b = safetensors.numpy.save({"a":np.array(range(6), dtype='<u4')})
print(a)
print(b)
```

```python
b'7\x00\x00\x00\x00\x00\x00\x00{"a":{"dtype":"U32","shape":[6],"data_offsets":[0,24]}}\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x02\x00\x00\x00\x03\x00\x00\x00\x04\x00\x00\x00\x05'
b'7\x00\x00\x00\x00\x00\x00\x00{"a":{"dtype":"U32","shape":[6],"data_offsets":[0,24]}}\x00\x00\x00\x00\x01\x00\x00\x00\x02\x00\x00\x00\x03\x00\x00\x00\x04\x00\x00\x00\x05\x00\x00\x00'
```

**Comment by Narsil** [comment]
No, this is by design, everything is supposed to be little endian.  https://github.com/huggingface/safetensors#notes.

We should definitely raise an error if it's coming as big-endian, but I deferred that code because I want to know if there are still big endian users out there and if there's any legit use case for trying to keep both.

IMO supporting both (like Fortran order) causes pain because now every reader has to wonder about what endianness is under the hood and paying the cost of conversion eventually. 

The current assumption is that almost all machines relevant for machine learning are using little-endian, so we can happily discard big-endian altogether and force those machines to write as little-endian, and they only have to pay the cost.

Since it's a big decision and I don't claim I know every kind of hardware out there, I delayed taking that relatively firm stance for now, I am waiting to be proven wrong :) .

**Comment by KOLANICH** [comment]
>The current assumption is that almost all machines relevant for machine learning are using little-endian

I know.

>so we can happily discard big-endian altogether and force those machines to write as little-endian, and they only have to pay the cost.

But currently it is not done. Instead it just dumps arrays from memory as they are.

>Since it's a big decision and I don't claim I know every kind of hardware out there, I delayed taking that relatively firm stance for now, I am waiting to be proven wrong :) .

At least the measures should be taken to enforce that all the arrays serialized now will be serialized and parsed as little endian. Because it would be very inconvenient if the format finds adoption and it turned out that some models were serialized as big endian, and we don't know which ones. Restricting serialization and parsing to le **right now** will not change the blob format and won't affect the most of hardware.

**Comment by Narsil** [comment]
Yup, you are perfectly correct ! 

As I said, we need to enforce little endianness. I just created a PR for that.
Thank you for the report.
